doc/PLUGINS.md

```diff
@@ -514,14 +514,49 @@ It is currently extremely restricted:
    commands, as these may become intermingled and break rule #1.
 3. the hook will be called before your plugin is initialized!
 
+This hook, unlike all the other hooks, is also strongly synchronous:
+`lightningd` will stop almost all the other processing until this
+hook responds.
+
 ```json
 {
+  "data_version": 42,
   "writes": [
     "PRAGMA foreign_keys = ON"
   ]
 }
 ```
 
+This hook is intended for creating continuous backups.
+The intent is that your backup plugin maintains two files:
+(1) an (obsolete) copy of the database, and (2) a file of
+database queries that will bring that database up-to-date.
+
+`data_version` is an unsigned 32-bit number that will always
+increment by 1 each time `db_write` is called.
+Your plugin **MUST** validate that the `data_version` is exactly 1
+higher than the previous `data_version` it was called.
+If it is not exactly that, some database queries might have gotten
+lost.
```

Comment:
We need to check for `n-1` (all good, we are in sync) and `n-2` (last transaction was not committed). If we are any other number we need to abort and kill `lightningd` (see [this logic](https://github.com/lightningd/plugins/blob/3ed5d935feafc842c702b4fa18c725d7f9311dab/backup/backup.py#L124-L155) for comparison).

Reply:
That seems to point to some extra behavior that should be documented as well (otherwise your backup plugin is depending on undocumented behavior...).  Do you mean that for every update made by "normal" `lightningd` operation, we increment twice, once for the update, another for its transaction being committed?

Reply:
No, each DB transaction bumps the number only once, but we can recover automatically from two cases:

 - Backup is at `n-1` and `db_write` tells us about `n`: this is the best case, we are in sync, the last transaction we wrote to the backup was successfully committed to the database, and we are continuing normally.
 - Backup is at `n-1` and `db_write` tells us about `n-1`: we were told about a DB transaction (`n-1`) during a prior run, but the DB transaction was not committed (remember we get told after the last statement, but before the tx gets committed). This can happen due to a crash, the DB not being ready, or a user killing `lightningd`. In this case we should roll back our `n-1`th transaction and apply the one we were just told about.

So really all we care about is that `n-1`, n-2`, and `n` are distinct from each other, and that there is no prior version that could be mistaken for any of these.

If we wanted to exaggerate we could also create a hash-chain: each DB transaction gets hashed, and hashes get accumulated over the entire lifetime of the DB. Thus each hash represents a unique state in the DB, and by remembering the last couple of hashes we can uniquely identify where we are in the backup. The numeric value of `data_version` is nothing more than a short-hand for that hash :wink:

The important thing remains that we can only recover automatically in one of the two situations above.

Reply:
Ah, thanks for the clarification, will add this text as well.